aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/MutationProcessor.java

```diff
@@ -115,13 +115,17 @@ private Completable drainMutationOutbox() {
      * @return A Completable that emits success when the item is processed, emits failure, otherwise
      */
     private <T extends Model> Completable processOutboxItem(PendingMutation<T> mutationOutboxItem) {
-        // First, publish the mutation over the network.
-        return publishToNetwork(mutationOutboxItem)
-            .flatMapCompletable(mutation ->
-                // Once the server knows about it, it's safe to remove from the outbox.
-                // This is done before merging, because the merger will refuse to merge
-                // if there are outstanding mutations in the outbox.
-                mutationOutbox.remove(mutationOutboxItem).andThen(merger.merge(mutation))
+        // First, mark the item as in-flight.
+        return mutationOutbox.markInFlight(mutationOutboxItem.getMutationId())
+            // Then, put it "into flight"
+            .andThen(publishToNetwork(mutationOutboxItem)
+                .flatMapCompletable(modelWithMetadata ->
+                    // Once the server knows about it, it's safe to remove from the outbox.
+                    // This is done before merging, because the merger will refuse to merge
+                    // if there are outstanding mutations in the outbox.
+                    mutationOutbox.remove(mutationOutboxItem.getMutationId())
```

Comment:
Quick question - if somehow there's a failure between the server getting it and this remove from outbox step, will it gracefully ignore the duplicate transaction next time the outbox attempts to perform it?

Reply:
My suspicion is if we get into this state, the outbox will be indefinitely wedged. We have some work to do here to make the outbox a) concurrent b) resilient

Reply:
I would think with the conditions on Updates and Deletes (or ID check on Create) it'd be fairly simple to avoid this being a problem if we aren't already. But yeah definitely not a launch blocker if we aren't since it's an edge case we could follow up with a fix for.